### ISSUE TITLE
addedToStage event fires before element is on stage

### ISSUE THREAD

**Issue by iamdustan** [uxebu/bonsai#128]
I’m in a situation where I need to add some meta data to the svg element drawn by Bonsai.

My current process involves something like the following:

``` javascript
// in runner
var layer = MyOwnObject;

// layer.element instanceof DisplayObject
layer.element.on('addedToStage'), function () {
  stage.sendMessage('addedToStage', layer);
})

// in DOM
stage = bonsai.run({})

stage.on('message:addedToStage', function (layer) {
  var bsId = '[data-bs-id="' + layer.element.id + '"]',
      el = stage.renderer.svg.root.querySelector(bsId);

  el.setAttribute('data-type', layer.type); // ==> Uncaught TypeError: Cannot call method 'setAttribute' of null 
})
```

If I add a timeout of 10 milliseconds before `stage.sendMessage`, then everything works as expected.

**Comment by basecode** [comment]
I think you're right, the name is misleading. It should be "willAddToStage" or sth similar. The process of adding a DisplayObject to the stage is async and at that point where this event is fired we actually can't promise that the DisplayObject was added.

It seems you actually wouldn't need to solve that problem in the renderer (DOM) if you would be able to create a custom "data-type" via the runner. I had a similar issue and solved it by extending the SVG Renderer and DisplayObject. See https://github.com/uxebu/bonsai/compare/master...attr-data. Unfortunately I'm not sure if this branch is ever going to be merged.

**Comment by iamdustan** [comment]
I had a conversation with @nonken on IRC the other day about the possibility of adding custom attributes to an element. That would make a lot of the things I’m trying to do much more straightforward, but the issue seems to be how will that work with a non-DOM based renderer?

**Comment by basecode** [comment]
> but the issue seems to be how will that work with a non-DOM based renderer?

Yes, that's the question. How could someone benefit of "meta-data" when we draw everything in a Canvas2D graphics context for example.

One thing about the `addedToStage` event I wasn't aware of when I answered your question yesterday is that the event works as expected. The reason for that is, that the _stage_ is an object on the runner side and the event is fired when a DisplayObject was added to that object. At the moment Bonsai doesn't provide an event that is fired when the DisplayObject was rendered.

**Comment by iamdustan** [comment]
Hmmm...any plans for firing an event when a DisplayObject is rendered? If only MutationObservers were more widely supported... :(
